crates/bevy_asset/src/utils.rs

```diff
@@ -0,0 +1,72 @@
+use concurrent_queue::ConcurrentQueue;
+use std::sync::Arc;
+
+pub(crate) struct EventQueue<T>(Arc<ConcurrentQueue<T>>);
+
+impl<T> EventQueue<T> {
+    pub fn new() -> Self {
+        Self(Arc::new(ConcurrentQueue::unbounded()))
+    }
+
+    pub fn sender(&self) -> EventSender<T> {
+        EventSender(self.0.clone())
+    }
+
+    pub fn receiver(&self) -> EventReceiver<T> {
+        EventReceiver(self.0.clone())
+    }
+
+    pub fn send(&self, value: T) {
+        self.0
+            .push(value)
+            .unwrap_or_else(|_| panic!("Failed to send value."));
+    }
+
+    pub fn try_recv(&self) -> Result<T, concurrent_queue::PopError> {
+        self.0.pop()
+    }
+
+    pub fn try_iter(&self) -> concurrent_queue::TryIter<T> {
+        self.0.try_iter()
+    }
+}
+
+impl<T> Clone for EventQueue<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+/// A strictly non-blocking sender for a multi-producer multi-consumer channel.
+#[derive(Debug)]
+pub struct EventSender<T>(Arc<ConcurrentQueue<T>>);
```

Comment:
Can we make this private? I'd prefer not to expose a reimplementation of a channel